R/table-name.R

```diff
@@ -0,0 +1,189 @@
+# table source = table id or sql
+
+# table id =
+# * interface to outside world; many ways to specify
+# * always refers to exactly one table
+# * but all converted to table name ASAP
+
+# table name
+# * escaped string
+# * internal (and backend) use only; not user facing
+# * can be vector of multiple names
+# * object names are always assumed to be table names
+
+# table_name --------------------------------------------------------------
+
+table_name <- function(x) {
+  check_character(x)
+  structure(x, class = c("dbplyr_table_name", "character"))
+}
+
+# So you can do SQL(table_name("foo"))
+setOldClass(c("dbplyr_table_name", "character"))
+
+is_table_name <- function(x) {
+  inherits(x, "dbplyr_table_name")
+}
+
+#' @export
+print.dbplyr_table_name <- function(x, ...) {
+  cat("<table_name> ", paste0(style_kw(x), collapse = ", "), "\n", sep = "")
+}
+
+#' @export
+`[.dbplyr_table_name` <- function(x, ...) {
+  table_name(NextMethod())
+}
+#' @export
+`[[.dbplyr_table_name` <- function(x, ...) {
+  table_name(NextMethod())
+}
+
+#' @export
+`c.dbplyr_table_name` <- function(x, ...) {
+  table_name(NextMethod())
+}
+
+make_table_name <- function(x, con, collapse = TRUE) {
+  needs_quote <- !vapply(x, function(x) inherits(x, "AsIs"), logical(1))
+  x[needs_quote] <- sql_escape_ident(con, x[needs_quote])
+  if (collapse) {
+    x <- paste0(x, collapse = ".")
+  }
+
+  table_name(x)
+}
+
+as_table_names <- function(x, con) {
+  make_table_name(x, con, collapse = FALSE)
+}
+
+# Extract just the table name from a full identifier
+table_name_table <- function(x, con) {
+  x <- as_table_name(x, con)
+
+  vapply(x, FUN.VALUE = character(1), function(x) {
+    if (x == "") return("")
+
+    out <- table_name_components(x, con)
+    out[[length(out)]]
+  })
+}
+table_name_components <- function(x, con) {
+  quote_char <- substr(sql_escape_ident(con, ""), 1, 1)
```

Comment:
This only works with the escaping dbplyr does. In some databases (e.g. SQLite) you can also escape with square brackets e.g. `SELECT * FROM [my weird table]`, see https://sqlite.org/lang_keywords.html (and now I just learned there are three ways to escape identifiers in SQLite...).

Reply:
Yeah, SQL server also supports escaping with `[]`. But I _think_ this should be ok in general, and it's still better than what we were doing before when folks supplied literal SQL as identifiers. We may need to make more changes here in the future, but I think it's a decent start.